helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -104,16 +104,17 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
       }
       return isValid;
     }).collect(Collectors.toList());
+
     if (candidateNodes.isEmpty()) {
       optimalAssignment.recordAssignmentFailure(replica,
           Maps.transformValues(hardConstraintFailures, this::convertFailureReasons));
       return Optional.empty();
     }
 
-    Function<AssignableNode, Double> calculatePoints =
-        (candidateNode) -> getAssignmentNormalizedScore(candidateNode, replica, clusterContext);
-
-    return candidateNodes.stream().max(Comparator.comparing(calculatePoints));
```

Comment:
can you try `candidateNodes.parallelStream().max(Comparator.comparing(calculatePoints));`?

Reply:
I tried, the current one is the fastest speed that I can get.
I think if you compare multiple times, the comparator will do calculating again and again. So pre-calculate the result is much faster somehow.

FYI, with your change:
generateClusterModel Took: 302
rebalanceAlgorithm.calculate took 9399
Finish calculating an assignment. Took: 9701
generateClusterModel Took: 122
rebalanceAlgorithm.calculate took 8546
Finish calculating an assignment. Took: 8668

With my change:
generateClusterModel Took: 295
rebalanceAlgorithm.calculate took 6520
Finish calculating an assignment. Took: 6815
generateClusterModel Took: 115
rebalanceAlgorithm.calculate took 6123
Finish calculating an assignment. Took: 6238